Fixes in monitoring, enable monitoring in Embedded GlassFish

OndroMih opened the pull request:
Ensures that the reported current thread count and busy thread count is exactly the number of threads reported by the pool. Even if monitoring is started while thread pool is actively processing requests.

A few improvements in Embedded GlassFish to enable monitoring and access it via JMX (it was completely broken before):

* if the flashlight-agent is already loaded, don't attempt to load it - that would fail on newer Java versions
* concatenate `probe-provider-class-names` MANIFEST elements from dependencies, using gmaven plugin script at build time. It references all probes that should be activated

Embedded GlassFish requires running with flashlight-agent on command line with -javaagent:flashlight-agent.jar on newer Java versions as it doesn't contain the agent. Dynamic loading of agent is not a good practice with newer Java versions anyway.

When started with command "enable-monitoring --mbean true", metric will be available via JMX beans:

Example to start Embedded GlassFish with thread pool and http service metrics:

```
java -javaagent:/path/to/flashlight-agent.jar -jar glassfish-embedded-all.jar 'enable-monitoring --modules thread-pool:http-service'
```

Or, alternatively, using the following properties in `glassfish.properties` in the current directory:

```
configs.config.server-config.monitoring-service.module-monitoring-levels.thread-pool=HIGH
configs.config.server-config.monitoring-service.module-monitoring-levels.http-service=HIGH
```

And command:

```
java -javaagent:/path/to/flashlight-agent.jar -jar glassfish-embedded-all.jar
```

Then in VisualVM (with MBeans extension), in the `amx` node:

<img width="1197" height="484" alt="image" src="https://github.com/user-attachments/assets/54b0720d-e225-48d1-bd3b-9277c1f1ce8d" />

If you don't see the `amx` node, you need to execute the `bootAMX` operation from VisualVM first (the added test demonstrates how to do that in code):

<img width="827" height="251" alt="image" src="https://github.com/user-attachments/assets/5803a773-ffe8-4907-99d3-4448a40af6c8" />

OndroMih committed:
IT Tests to cover thread pool monitoring. Some failing.

OndroMih committed:
Thread pool monitoring - set number of threads directly from pool info

OndroMih committed:
More IT Tests to cover thread pool monitoring.

OndroMih committed:
Set setCurrentThreadCountEvent directly

OndroMih committed:
Fix monitoring in Embedded GlassFish

Requires running with flashlight-agent on command line with -javaagent:flashlight-agent.jar
When started with command "enable-monitoring --mbean true", metric will be available via JMX beans

OndroMih committed:
Add missing add-exports 

Needed for monitoring, at least in Embedded GlassFish. Otherwise:

java.lang.reflect.InaccessibleObjectException: Unable to make public java.util.Enumeration jdk.internal.loader.BuiltinClassLoader.findResources(java.lang.String) throws java.io.IOException accessible: module java.base does not "exports jdk.internal.loader" to unnamed module @1de0aca6
	at java.base/java.lang.reflect.AccessibleObject.throwInaccessibleObjectException(AccessibleObject.java:391)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:367)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:315)
	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:203)
	at java.base/java.lang.reflect.Method.setAccessible(Method.java:197)
	at com.sun.enterprise.module.single.ManifestProxy.<init>(ManifestProxy.java:71)
	at com.sun.enterprise.module.single.ProxyModuleDefinition.generate(ProxyModuleDefinition.java:146)
	at com.sun.enterprise.module.single.ProxyModuleDefinition.getManifest(ProxyModuleDefinition.java:134)
	at org.glassfish.admin.monitor.MonitoringBootstrap.addProvider(MonitoringBootstrap.java:298)

OndroMih committed:
Refactor flashlight-agent attachment code to simplify it.

OndroMih committed:
Document monitoring for Embedded GlassFish

OndroMih committed:
Test for monitoring of Embedded GlassFish via JMX

OndroMih committed:
Test improvement - connect to JMX via socket instead of port

Disable test for Embedded Web - it doesn't support AMX because it doesn't include glassfish-mbeanserver artifact.

OndroMih committed:
Always use Hamcrest assertions

OndroMih committed:
Fix the testThreadPoolSizeCycling test

OndroMih committed:
Fixes in thread pool monitoring

- do not unregister probes when deleting listener, unless it's the last listener
- count busy threads in the monitor and pass info to stats (maybe not necessary)
- catch out of bounds exception in probes, which sometimes happens, probably because of a probe is removed concurrently from the array while iterating through the collection
- global stats updates with the setCurrentThreadCountEvent event

dmatej commented on nucleus/flashlight/framework/src/main/java/org/glassfish/flashlight/provider/FlashlightProbe.java:
```diff
@@ -148,11 +150,15 @@ public void fireProbe(Object[] params) {
+                }
             }
+        } catch (IndexOutOfBoundsException e) {
+            logger.log(FINE, "", e);
```
Really? Maybe it would be better to clone the list and then use just foreach.

dmatej commented on appserver/extras/embedded/pom.xml:
```diff
@@ -52,4 +56,62 @@
+            <plugins>
+                <plugin>
+                    <groupId>org.codehaus.gmaven</groupId>
+                    <artifactId>groovy-maven-plugin</artifactId>
```
I would prefer using java and exec plugin; not sure if it is doable. 
The groovy is not so well readable for us and we tend to do evil things: catching exceptions without processing them, closing jars just when their processing doesn't throw exceptions instead of try-with ...

OndroMih committed:
Improve iterating invokers in flashlight

OndroMih committed:
Reset thread pool monitoring stats to current values instead of 0

Also adds tests for it.

Also adds support for a system property glassfish.remote, to run tests against a remote server instead of setting up a local (managed) one and stopping it after tests. This helps in debugging test scenarios.

OndroMih committed:
Fix race conditions in thread pool monitoring

Introduced by a previous "fix" :)

pzygielo commented on appserver/extras/embedded/pom.xml:
```diff
@@ -52,4 +56,62 @@
+            <plugins>
+                <plugin>
+                    <groupId>org.codehaus.gmaven</groupId>
+                    <artifactId>groovy-maven-plugin</artifactId>
```
What is the version of the plugin used here?

pzygielo replied:
c7d4a9b73f965d96ca7fbd5d009af9717908b305

OndroMih committed:
Set version of the gmaven plugin to the currently latest version

pzygielo commented on .github/workflows/build-linux.yml:
```diff
@@ -28,3 +28,8 @@ jobs:
         retention-days: 3
+    - name: Publish Test Report
+      uses: mikepenz/action-junit-report@v5
+      if: success() || failure() # always run even if the previous step fails
```
Why not `always()` and no comment? (I find the current comment as misleading a bit as it says _always_, but `cancelled` is not covered...)

pzygielo commented on appserver/extras/embedded/all/pom.xml:
```diff
@@ -127,6 +127,17 @@
                 </executions>
             </plugin>
 
+            <plugin>
```
If it's only me - let's keep it the way it is now.
Thanks.

OndroMih commented:
It seems that each time when the Windows build got cancelled after 6 hours, it was stuck in the MonitoringTest for Embedded GlassFish, which starts GlassFish with the flashlight agent. For now, I disabled that test on WIndows. In the future, we want to refactor the monitoring engine and get rid of the agent anyway.

OndroMih commented on .github/workflows/build-linux.yml:
```diff
@@ -3,6 +3,8 @@ name: Build GlassFish on Ubuntu
 on:
   pull_request:
     branches: [ "master", "7.0", "8.0" ]
+permissions:
```
I think @dmatej you reverted too much in this commit: https://github.com/eclipse-ee4j/glassfish/pull/25752/commits/3fb9a43c0f9a5f8f91a95826d81fd820dda8b564

I added 
```
permissions:
  checks: write
```
for the `action-junit-report` step, although I added it with a commit that only mentions adding RAM disk. Though it seems that have this permission block doesn't have any effect.

I don't understand how the `action-junit-report` exactly works. Now I tried to enable annotation-only mode, hopefully it will provide us what I seek for - information about which tests fail if some tests fail. Right now, this step only provides the number of passed and failed tests.

dmatej replied:
It is a work in progress, I will go over those commits later one by one and move some changes to own PR. First I need to understand what is really happening and why the build freezes just on GHA.

dmatej committed:
Quick cleanup and minor changes in AMX and mbeanserver code

- Partially removed printing stacktraces when we throw after that
- AMXConfigLoader - refactoring
  - condition was always true, and we could move some code to constructor
    to make it more visible
  - synchronization was not useful here, the code is synchronized
    from the outside too
- Formatting, blank lines, annotations
- Throwing "unreachable" exceptions instead of swallowing them

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej commented:
Status: I have rebased #25768 to this, fixed conflicts (I made them). Now I will create another PR just with Ondro's changes in GHA, then I will rebase again, I will try yet to improve the performance of those new tests in application-tests, so we could avoid those tags. I have a new weapon for that, learned new usages of the Stream API :-)

Otherwise, nothing new about Jenkins performance, unfortunately.

dmatej committed:
Code cleanup in embedded runnable tests

- Special matcher
- Deleting temporary files
- Fixed Windows STDERR processing - we must read the output

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed resource leak in CollectLogFiles

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Added onSpinWait to waitFor and removed InterruptedException

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
ThreadPoolMonitoringTest - reactoring, fixes, stabilization

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Processed review comments - removed fixme, renamed class

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Better control over threads and locks both on client and server side

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Added waitForTaskCountStoppedChanging

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed test condition

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>
